### ISSUE TITLE
Working with internal proxy

### ISSUE THREAD

**Issue by Aokrip** [siamaksade/openshift-cd-demo#26]
Hi,

Do you know if your demo works with an internal proxy to access to the internet ?

Currently, when I'm trying to run the build from Jenkins, the console output is just 

```
Started by user anonymous
[Pipeline] Allocate node : Start
Still waiting to schedule task
Waiting for next available executor
```

No slave pod is created on my OpenShift project
So, I had a look to the Jenkins pod logs, and every ten seconds during the build, the following error is triggered :

```
Sep 09, 2016 3:15:54 AM org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud provision
INFO: Excess workload after pending Spot instances: 1
Sep 09, 2016 3:15:54 AM org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud provision
WARNING: Failed to count the # of live instances on Kubernetes
java.lang.IllegalArgumentException: No enum constant com.ning.http.client.ProxyServer.Protocol.http
    at java.lang.Enum.valueOf(Enum.java:238)
    at com.ning.http.client.ProxyServer$Protocol.valueOf(ProxyServer.java:33)
    at io.fabric8.kubernetes.client.BaseClient.createHttpClient(BaseClient.java:152)
    at io.fabric8.kubernetes.client.BaseClient.<init>(BaseClient.java:68)
    at io.fabric8.kubernetes.client.DefaultKubernetesClient.<init>(DefaultKubernetesClient.java:89)
    at org.csanchez.jenkins.plugins.kubernetes.KubernetesFactoryAdapter.createClient(KubernetesFactoryAdapter.java:71)
    at org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud.connect(KubernetesCloud.java:197)
    at org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud.addProvisionedSlave(KubernetesCloud.java:415)
    at org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud.provision(KubernetesCloud.java:297)
    at hudson.slaves.NodeProvisioner$StandardStrategyImpl.apply(NodeProvisioner.java:700)
    at hudson.slaves.NodeProvisioner.update(NodeProvisioner.java:305)
    at hudson.slaves.NodeProvisioner.access$000(NodeProvisioner.java:58)
    at hudson.slaves.NodeProvisioner$NodeProvisionerInvoker.doRun(NodeProvisioner.java:797)
    at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:50)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)
```

I tried to figure out how to set a proxy on Jenkins, unfortunately nothing worked at the moment.
Do you have an idea ?

Thank you
Best regards,

Cyril

**Comment by dpalmeira** [comment]
Hi cyril-dussert, did you resolve the issue? I have a similar problem :S

**Comment by Aokrip** [comment]
Hi,

I managed to resolve the issue (forgot to update :x) It seems that a simple update of the Kubernetes plugin on jenkins resolve the issue (don't forget to setup proxy setting for checking plugin updates)

**Comment by dpalmeira** [comment]
Just a question, currently I have the 0.8 version (it comes with the image), if you see the official repo, this is the latest version: http://updates.jenkins-ci.org/download/plugins/kubernetes/

I found there is a new version in the github repo:
https://github.com/jenkinsci/kubernetes-plugin/tree/kubernetes-0.9

What is the version are you using?

thx for your help I'm getting crazy with this :)

**Comment by dpalmeira** [comment]
I just downloaded, compiled and deployed the kubernetes plugin v0.9, restart my master Jenkins (recreate pod), and same result. It fails to connect to kubernetes api. (plugin version now shows now 0.9-SNAPSHOT).

I've saw, my problem is different than yours.

I'm following this tutorial:
https://blog.openshift.com/openshift-3-2-jenkins-s2i-slave-pods/

**Originally there are 2 errors times. One during the startup of master Jenkins image:**

INFO: Listed all plugins
Oct 25, 2016 7:08:08 AM jenkins.InitReactorRunner$1 onAttained
INFO: Prepared all plugins
Oct 25, 2016 7:08:08 AM jenkins.InitReactorRunner$1 onAttained
INFO: Started all plugins
Oct 25, 2016 7:08:13 AM hudson.ExtensionFinder$GuiceFinder$FaultTolerantScope$1 error
WARNING: Failed to instantiate Key[type=io.fabric8.jenkins.openshiftsync.GlobalPluginConfiguration, annotation=[none]]; skipping this component
com.google.inject.ProvisionException: Guice provision errors:

1) Error injecting constructor, io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
  at io.fabric8.jenkins.openshiftsync.GlobalPluginConfiguration.<init>(GlobalPluginConfiguration.java:52)

1 error
    at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:52)
    at com.google.inject.Scopes$1$1.get(Scopes.java:65)
    at hudson.ExtensionFinder$GuiceFinder$FaultTolerantScope$1.get(ExtensionFinder.java:424)
    at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:41)
    at com.google.inject.internal.InjectorImpl$3$1.call(InjectorImpl.java:1005)
    at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1051)
    at com.google.inject.internal.InjectorImpl$3.get(InjectorImpl.java:1001)
    at hudson.ExtensionFinder$GuiceFinder._find(ExtensionFinder.java:386)
    at hudson.ExtensionFinder$GuiceFinder.find(ExtensionFinder.java:377)
    at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:389)
    at hudson.ExtensionList.load(ExtensionList.java:349)
    at hudson.ExtensionList.ensureLoaded(ExtensionList.java:287)
    at hudson.ExtensionList.iterator(ExtensionList.java:156)
    at jenkins.model.Jenkins.getDescriptorByType(Jenkins.java:1233)
    at hudson.plugins.git.GitSCM.onLoaded(GitSCM.java:1732)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:106)
    at hudson.init.TaskMethodFinder$TaskImpl.run(TaskMethodFinder.java:176)
    at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:282)
    at jenkins.model.Jenkins$8.runTask(Jenkins.java:926)
    at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:210)
    at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
    at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:57)
    at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:418)
    at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:58)
    at io.fabric8.jenkins.openshiftsync.BuildConfigWatcher.start(BuildConfigWatcher.java:65)
    at io.fabric8.jenkins.openshiftsync.GlobalPluginConfiguration.configChange(GlobalPluginConfiguration.java:117)
    at io.fabric8.jenkins.openshiftsync.GlobalPluginConfiguration.<init>(GlobalPluginConfiguration.java:54)
    at io.fabric8.jenkins.openshiftsync.GlobalPluginConfiguration$$FastClassByGuice$$7bc306ea.newInstance(<generated>)
    at com.google.inject.internal.cglib.reflect.$FastConstructor.newInstance(FastConstructor.java:40)
    at com.google.inject.internal.DefaultConstructionProxyFactory$1.newInstance(DefaultConstructionProxyFactory.java:61)
    at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:108)
    at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:88)
    at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:269)
    at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46)
    at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1058)
    at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
    ... 27 more
**Caused by: java.io.IOException: Unexpected response code for CONNECT: 503**
    at okhttp3.internal.connection.RealConnection.createTunnel(RealConnection.java:329)
    at okhttp3.internal.connection.RealConnection.buildTunneledConnection(RealConnection.java:155)
    at okhttp3.internal.connection.RealConnection.connect(Real

And another one, when I'm trying to build the example task:

INFO: Excess workload after pending Spot instances: 1
Oct 25, 2016 7:13:44 AM org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud provision
WARNING: Failed to count the # of live instances on Kubernetes
io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
    at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:57)
    at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:418)
    at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:58)
    at org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud.addProvisionedSlave(KubernetesCloud.java:511)
    at org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud.provision(KubernetesCloud.java:392)
    at hudson.slaves.NodeProvisioner$StandardStrategyImpl.apply(NodeProvisioner.java:700)
    at hudson.slaves.NodeProvisioner.update(NodeProvisioner.java:305)
    at hudson.slaves.NodeProvisioner.access$000(NodeProvisioner.java:58)
    at hudson.slaves.NodeProvisioner$NodeProvisionerInvoker.doRun(NodeProvisioner.java:797)
    at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:50)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
**Caused by: java.net.SocketTimeoutException: timeout**
    at okio.Okio$3.newTimeoutException(Okio.java:207)
    at okio.AsyncTimeout.exit(AsyncTimeout.java:261)
    at okio.AsyncTimeout$2.read(AsyncTimeout.java:215)
    at okio.RealBufferedSource.indexOf(RealBufferedSource.java:306)
    at okio.RealBufferedSource.indexOf(RealBufferedSource.java:300)
    at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:196)
    at com.squareup.okhttp.internal.http.Http1xStream.readResponse(Http1xStream.java:186)
    at com.squareup.okhttp.internal.io.RealConnection.createTunnel(RealConnection.java:270)
    at com.squareup.okhttp.internal.io.RealConnection.connectTls(RealConnection.java:172)
    at com.squareup.okhttp.internal.io.RealConnection.connectSocket(RealConnection.java:149)
    at com.squareup.okhttp.internal.io.RealConnection.connect(RealConnection.java:112)
    at com.squareup.okhttp.internal.http.StreamAllocation.findConnection(StreamAllocation.java:184)
    at com.squareup.okhttp.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:126)
    at com.squareup.okhttp.internal.http.StreamAllocation.newStream(StreamAllocation.java:95)
    at com.squareup.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:281)
    at com.squareup.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:224)
    at com.squareup.okhttp.Call.getResponse(Call.java:286)
    at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:243)
    at io.fabric8.kubernetes.client.utils.HttpClientUtils$3.intercept(HttpClientUtils.java:93)
    at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:232)
    at com.squareup.okhttp.Call.getResponseWithInterceptorChain(Call.java:205)
    at com.squareup.okhttp.Call.execute(Call.java:80)
    at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:230)
    at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:416)
    ... 15 more
**Caused by: java.net.SocketException: Socket closed**
    at java.net.SocketInputStream.read(SocketInputStream.java:203)
    at java.net.SocketInputStream.read(SocketInputStream.java:141)
    at okio.Okio$2.read(Okio.java:139)
    at okio.AsyncTimeout$2.read(AsyncTimeout.java:211)
    ... 36 more

Any idea?
Could that be caused by some missing configuration in the proxies?

thx,

**Comment by dpalmeira** [comment]
I revolved the first part of the connection. It was caused by the wrong URL to point the kubernetes api. :)

**Comment by dpalmeira** [comment]
Now, I have a new issue, the slave cannot connect to the Jenkins master on port 50000
I'm investigating it.

**Comment by dpalmeira** [comment]
Solved!, the problem were two:
1) The Jenkins URL with IP didn't work for me I had to put the URL by doing the following format:
<service>.<project>.svc.cluster.local
e.g. Jenkins URL field, enter http://jenkins.jenkins.svc.cluster.local:8080
Generally the .svc.cluster.local is already excluded from your proxy.

After that, the jenkins-slave was able to download the remoting.jar and run it using java command.

2) The second problem was jnlp connection to port 50000 didn't work with 503 Service Unavailable.
It was causes by proxy env http_proxy. I found in some forum jnlp is not using JAVA_OPTS settings and in my case it was trying the connection through the proxy.

In order to resolve it, I had to modify the original script included in my example corresponding to the slave part: run-jnlp-client to unset the http_proxy variable and restore it after java call like this:

```
  **# Disabled proxy cause jnlp use the env variable and does not use the $JAVA_OPTS
  echo "Original values of http_proxy=${http_proxy}"
  BK_HTTP_PROXY=$http_proxy
  unset http_proxy**

  echo Running java $JAVA_OPTS -cp $JAR hudson.remoting.jnlp.Main -headless $PARAMS "$@"
  cd ${JENKINS_DIR} && exec java $JAVA_OPTS \
    -cp $JAR hudson.remoting.jnlp.Main -headless $PARAMS "$@"

  **# Restore http_proxy values
  BK_HTTP_PROXY=${http_proxy}
  export http_proxy=${BK_HTTP_PROXY}
  echo "After restore values of http_proxy=${http_proxy}"**
```

With this in place, the jenkins-slave connected successfully to the jenkins-master under Openshift 3.2 in my company.

I hope this helps to someone with the same problem.
